### PR TITLE
refactor: optimize the code of angle calculations

### PR DESCRIPTION
**Pull request by GuoBinyong**
1. Improved computing performance
2. Superior annotation

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14519/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14519/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14519/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14519/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14519/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14519/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14519/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14519/merge#BCU1XR#0

**Review by Popov72 — APPROVED** [review]

File: packages/dev/core/src/Maths/math.path.ts [hunk 1]
```diff
@@ -96,17 +96,18 @@ export class Angle {
...
-        const na = a.normalizeToNew() as Vec;
-        const nb = b.normalizeToNew() as Vec;
-        let cosv = na.dot(nb as any);
-        cosv = Scalar.Clamp(cosv, -1, 1);
-        const theta = Math.acos(cosv);
-        return new Angle(theta);
+        let product = a.lengthSquared() * b.lengthSquared();
+        if (product === 0) return new Angle(Math.PI / 2);
+        product = Math.sqrt(product);
+        let cosVal = (a.constructor as any).Dot(b) / product;
```
Comment (RaananW) [inline packages/dev/core/src/Maths/math.path.ts:107]: is `any` needed here? dont all Vec types have this method defined?

**Review by RaananW — COMMENTED** [review]
Awesome!

A small question regarding typing, and a general question - you say the performance is better. how did you test that?

File: packages/dev/core/src/Maths/math.path.ts (hunk 1 above)
Comment (GuoBinyong) [reply packages/dev/core/src/Maths/math.path.ts]: This was written before my pr passed, and not all vectors had dot methods at the time.

Now I've changed it around

**Comment by GuoBinyong** [comment]
> Awesome!
> 
> A small question regarding typing, and a general question - you say the performance is better. how did you test that?

In the case of Vector3, the following operations are reduced:
- 1 operation of 'Math.sqrt()' in 'normalizeToNew()'
- 1 division operation in 'normalizeToNew()'
- 6 times of multiplication in 'normalizeToNew()'
- 4 function calls

**Review by Popov72 — APPROVED** [review]

File: packages/dev/core/src/Maths/math.path.ts (hunk 1 above)
Comment (RaananW) [reply packages/dev/core/src/Maths/math.path.ts]: But still the any is left. That's what I'm trying to understand - why is it needed

**Comment by RaananW** [comment]
there is still an "any". i'll merge it and fix it later.

**Pull request merged**